File: georocket-server/src/main/java/io/georocket/index/xml/XalAddressIndexerFactory.java
package io.georocket.index.xml;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import io.georocket.query.KeyValueQueryPart;
import io.georocket.query.KeyValueQueryPart.ComparisonOperator;
import io.georocket.query.QueryPart;
import io.georocket.query.StringQueryPart;
import io.vertx.core.json.JsonObject;

import java.util.Map;

import static io.georocket.query.ElasticsearchQueryHelper.gtQuery;
import static io.georocket.query.ElasticsearchQueryHelper.gteQuery;
import static io.georocket.query.ElasticsearchQueryHelper.ltQuery;
import static io.georocket.query.ElasticsearchQueryHelper.lteQuery;
import static io.georocket.query.ElasticsearchQueryHelper.multiMatchQuery;
import static io.georocket.query.ElasticsearchQueryHelper.termQuery;

/**
 * Creates instances of {@link XalAddressIndexer}
 * @author Michel Kraemer
 */
public class XalAddressIndexerFactory implements XMLIndexerFactory {
  @Override
  public XMLIndexer createIndexer() {
    return new XalAddressIndexer();
  }

  @Override
  public Map<String, Object> getMapping() {
    return ImmutableMap.of("dynamic_templates", ImmutableList.of(ImmutableMap.of(
      "addressFields", ImmutableMap.of(
        "path_match", "address.*",
        "mapping", ImmutableMap.of(
          "type", "keyword"
        )
      )
    )));
  }

  @Override
  public MatchPriority getQueryPriority(QueryPart queryPart) {
    if (queryPart instanceof StringQueryPart ||
      queryPart instanceof KeyValueQueryPart) {
      return MatchPriority.SHOULD;
    }
    return MatchPriority.NONE;
  }

  @Override
  public JsonObject compileQuery(QueryPart queryPart) {
    if (queryPart instanceof StringQueryPart) {
      // match values of all fields regardless of their name
      String search = ((StringQueryPart)queryPart).getSearchString();
      return multiMatchQuery(search, "address.*");
    } else if (queryPart instanceof KeyValueQueryPart) {
      KeyValueQueryPart kvqp = (KeyValueQueryPart)queryPart;
      String key = kvqp.getKey();
      String value = kvqp.getValue();
      ComparisonOperator comp = kvqp.getComparisonOperator();
      String name = "address." + key;
      switch (comp) {
        case EQ:
          return termQuery(name, value);
        case GT:
          return gtQuery(name, value);
        case GTE:
          return gteQuery(name, value);
        case LT:
          return ltQuery(name, value);
        case LTE:
          return lteQuery(name, value);
      }
    }
    return null;
  }
}

File: georocket-server/src/main/kotlin/io/georocket/index/xml/XalAddressIndexerFactory.kt
package io.georocket.index.xml

import io.georocket.query.QueryCompiler.MatchPriority
import io.georocket.query.QueryPart
import io.georocket.query.QueryPart.ComparisonOperator
import io.georocket.query.StringQueryPart
import io.vertx.core.json.JsonObject
import io.vertx.kotlin.core.json.jsonObjectOf

/**
 * Creates instances of [XalAddressIndexer]
 * @author Michel Kraemer
 */
class XalAddressIndexerFactory : XMLIndexerFactory {
  override fun createIndexer() = XalAddressIndexer()

  override fun getQueryPriority(queryPart: QueryPart): MatchPriority {
    return when (queryPart) {
      is StringQueryPart -> MatchPriority.SHOULD
      else -> MatchPriority.NONE
    }
  }

  override fun compileQuery(queryPart: QueryPart): JsonObject? {
    return when (queryPart) {
      is StringQueryPart -> {
        val v: Any = when (queryPart.comparisonOperator) {
          null, ComparisonOperator.EQ -> queryPart.value
          ComparisonOperator.GT -> jsonObjectOf("\$gt" to queryPart.value)
          ComparisonOperator.GTE -> jsonObjectOf("\$gte" to queryPart.value)
          ComparisonOperator.LT -> jsonObjectOf("\$lt" to queryPart.value)
          ComparisonOperator.LTE -> jsonObjectOf("\$lte" to queryPart.value)
        }

        if (queryPart.key == null) {
          jsonObjectOf("\$or" to XalAddressIndexer.Companion.Keys.values().map { key ->
            jsonObjectOf("address.${key.key}" to v)
          })
        } else {
          jsonObjectOf("address.${queryPart.key}" to v)
        }
      }

      else -> null
    }
  }
}
